pkg/cluster/registerWithHive.go

```diff
@@ -0,0 +1,110 @@
+		return nil, nil
+	}
+	m.log.Info("collecting registration data from new cluster")
+
+	cv, err := m.configcli.ConfigV1().ClusterVersions().Get(ctx, "version", metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	clusterID := cv.Spec.ClusterID
+
+	// TODO: When hive support first party principles we'll need to send both first party and cluster service principles
+	servicePrincipal := azure.Credentials{
+		TenantID:       m.subscriptionDoc.Subscription.Properties.TenantID,
+		SubscriptionID: m.subscriptionDoc.ID,
+		ClientID:       m.doc.OpenShiftCluster.Properties.ServicePrincipalProfile.ClientID,
+		ClientSecret:   string(m.doc.OpenShiftCluster.Properties.ServicePrincipalProfile.ClientSecret),
+	}
+
+	spBytes, err := json.Marshal(servicePrincipal)
+	if err != nil {
+		return nil, err
+	}
+
+	kubeconfig := string(m.doc.OpenShiftCluster.Properties.AROServiceKubeconfig)
+
+	resourcegroupname := stringutils.LastTokenByte(m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID, '/')
+
+	hiveWorkloadCluster := &hiveregistrator.WorkloadCluster{
+		SubscriptionID:              m.subscriptionDoc.ID,
+		ClusterName:                 m.doc.OpenShiftCluster.Name,
+		ResourceGroupName:           resourcegroupname,
+		Location:                    m.doc.OpenShiftCluster.Location,
+		BaseDomain:                  m.env.Domain(),
+		BaseDomainResourceGroupName: m.env.ResourceGroup(),
+		InfraID:                     m.doc.OpenShiftCluster.Properties.InfraID,
+		ClusterID:                   string(clusterID),
+		KubeConfig:                  kubeconfig,
+		ServicePrincipal:            string(spBytes),
+	}
+
+	return hiveWorkloadCluster, nil
+}
+
+func (m *manager) registerWithHive(ctx context.Context, workloadCluster *hiveregistrator.WorkloadCluster) error {
+	// don't fail the installation on hive related errors
+	if m.hiveRegistrator == nil {
+		m.log.Info("no hive registrator, skipping hive registration")
+		return nil
+	}
+	m.log.Info("registering with hive")
+
+	cd, err := m.hiveRegistrator.Register(ctx, workloadCluster)
+	if err != nil {
+		return err
+	}
+
+	m.doc, err = m.db.PatchWithLease(ctx, m.doc.Key, func(doc *api.OpenShiftClusterDocument) error {
+		doc.OpenShiftCluster.Properties.HiveProfile = api.HiveProfile{
+			NamespaceName: cd.Namespace,
+		}
+		return nil
+	})
+
+	return err
+}
+
+func (m *manager) verifyRegistration(ctx context.Context) (bool, error) {
+	// Never fail
+	if m.hiveRegistrator == nil {
+		m.log.Info("no hive registrator, skipping cluster verification")
+		return true, nil
+	}
+
+	m.log.Info("verifing cluster registration in hive")
+
+	isConnected, reason, err := m.hiveRegistrator.IsConnected(ctx, m.doc.OpenShiftCluster.Properties.HiveProfile.NamespaceName)
+	if err != nil {
+		m.log.Info(err)
+		return true, nil
+	}
+
+	if !isConnected {
+		m.log.Printf("hive is not able to connect to cluster %s", reason)
+		return true, nil
+	}
+
+	m.log.Info("cluster adopted successfully")
+
+	return true, nil
+}
```

Comment:
I think this function in it's context needs some tweaking due to it being called in steps.Condition with a timeout of 2 minutes.

The changes below will do the following:
1. return immediately if hive doesn't have a registerator
2. if the hive registrator returns an error, it will log the error, but continue the installation
3. if the hive registrator does not return an error,  but it's not connected it will retry the function
4. else success

```suggestion
func (m *manager) verifyRegistration(ctx context.Context) (bool, error) {
	// Never fail
	if m.hiveRegistrator == nil {
		m.log.Info("no hive registrator, skipping cluster verification")
		return true, nil
	}

	m.log.Info("verifing cluster registration in hive")
	isConnected, reason, err := m.hiveRegistrator.IsConnected(ctx, m.doc.OpenShiftCluster.Properties.HiveProfile.NamespaceName)
	if err != nil {
		return false, err
	}

	if !isConnected {
		m.log.Printf("hive is not able to connect to cluster %s", reason)
		return false, nil
	}

	m.log.Info("cluster adopted successfully")
	return true, nil
}
```


steps.Conditions looks like the below, so if an error happens within the condition function, it will return it but not fail if the fail parameter is `false`.  If the function returns false, it will rerun it.  

```
	err := wait.PollImmediateUntil(pollInterval, func() (bool, error) {
		// We use the outer context, not the timeout context, as we do not want
		// to time out the condition function itself, only stop retrying once
		// timeoutCtx's timeout has fired.
		return c.f(ctx)
	}, timeoutCtx.Done())

	if err != nil && !c.fail {
		log.Warnf("step %s failed but has configured 'fail=%t'. Continuing. Error: %s", c, c.fail, err.Error())
		return nil
	}
	return err
```

Reply:
Given [this](https://github.com/Azure/ARO-RP/pull/2145#discussion_r913089133), I think we should only return `true` when cluster addopted. In all other cases we should either return an `false, err` (to stop immediately) or `false, nil` (to continue trying).

P.S.: `wait.PollImmediateUntil` stops immediately when there is an error (even if we return `false`).